xl2times/transforms.py

```diff
@@ -2218,6 +2219,99 @@ def _match_wildcards(
+    region: str | None,
+    year: int | None,
+) -> pd.Index:
+    qs = []
+    if process is not None:
+        qs.append(f"process in ['{process}']")
+    if commodity is not None:
+        qs.append(f"commodity in ['{commodity}']")
+    if attribute is not None:
+        qs.append(f"attribute == '{attribute}'")
+    if region is not None:
+        qs.append(f"region == '{region}'")
+    if year is not None:
+        qs.append(f"year == {year}")
+    return table.query(" and ".join(qs)).index
+
+
+def eval_and_update(table: DataFrame, rows_to_update: pd.Index, new_value: str) -> None:
+    """Performs an inplace update of rows `rows_to_update` of `table` with `new_value`,
+    which can be a update formula like `*2.3`.
+    """
+    if isinstance(new_value, str) and new_value[0] in {"*", "+", "-", "/"}:
+        old_values = table.loc[rows_to_update, "value"]
+        updated = old_values.astype(float).map(lambda x: eval("x" + new_value))
+        table.loc[rows_to_update, "value"] = updated
+    else:
+        table.loc[rows_to_update, "value"] = new_value
+
+
+def _eval_updates(eval_table: pd.DataFrame, updates: pd.DataFrame) -> pd.DataFrame:
+    """
+    Evaluate the update formulas (e.g. *,/,+,-) in updates on matching rows in table, returning updated rows.
+
+    This uses sqldf and an in-mem sqlite database to perform this complex join in a single operation and is significantly faster than row iteration
+    for large tables (which is necessary as this cannot be done in a single `df.query()` call due to the conditional matching.
+
+    1. For each for in updates, fFind rows from `table` matching the same Process, Commodity, Attribute, Region, and Year
+    2. Apply the update formula (copy the value or apply basic numerical operations) to the value of each matching row
+
+    Args:
+        eval_table:
+        updates:
+
+    Returns: the updated rows as a dataframe
+    """
+    new_tables = []
+    join_cols = ["process", "commodity", "attribute", "region", "year"]
+
+    # Join table with update.value by process, commodity, attribute, region, year, where a None value in a key column matches all values in the
+    # corresponding column of table. This is equivalent to the SQL (u.process = t.process OR u.process IS NULL) join conditions.
+    # For any columns values in update that are none, we match all values in the corresponding columns of table
+    joined = sqldf(
+        """
+        SELECT t.*, u.value AS 'value_upd', u.source_filename as 'source_filename_upd'
+        FROM updates AS u
+        LEFT JOIN eval_table AS t ON
+        (u.process = t.process OR u.process IS NULL) AND
+        (u.commodity = t.commodity OR u.commodity IS NULL) AND
+        (u.attribute = t.attribute OR u.attribute IS NULL) AND
+        (u.region = t.region OR u.region IS NULL) AND
+        (u.year = t.year OR u.year IS NULL)
+        WHERE t.value IS NOT NULL
+        """,
+        env={"updates": updates, "eval_table": eval_table},
+    )
```

Comment:
This `sqldf()` call replaces the old  `rows_to_update = query(...)` logic with a single vectorised call.